File: neetcode150/15_greedy/0055_jump_game.py
"""
=====================================================================
LeetCode 55 : Jump Game                                          (Medium)
https://leetcode.com/problems/jump-game/
Category   : Greedy
---------------------------------------------------------------------
PROBLEM
    Start at index 0 of `nums`. From index i you can jump at most
    nums[i] steps forward. Return True if you can reach the last
    index.
---------------------------------------------------------------------
INTUITION
    Only the FURTHEST reachable index matters. Track one scalar:
        reach = max(reach, i + nums[i])
    If the sweep pointer passes `reach`, that index is unreachable →
    False. Larger reach strictly dominates smaller (exchange
    argument), so this single-number greedy is exact.
---------------------------------------------------------------------
APPROACH 1 — Greedy furthest-reach (BEST)
    Time  Complexity : O(n)   Space Complexity : O(1)
---------------------------------------------------------------------
APPROACH 2 — DP frontier O(n²) (why greedy is better)
    reachable[i] = True and stamp every cell i+1 .. i+nums[i]. The
    reachable set is always ONE interval [0, reach] — the DP table
    is n redundant booleans where a single int suffices.
    Time  Complexity : O(n²)   Space Complexity : O(n)
---------------------------------------------------------------------
DEEP DIVE — exchange argument for furthest-reach
    Induction: greedy position after k moves ≥ optimal position after
    k moves. Since our current index is ≥ theirs, our jump options
    cover all of theirs, so we can imitate any winning path. This
    monotone-domination property is the fingerprint of a correct
    greedy: local "go furthest" is never worse than any rival.
---------------------------------------------------------------------
REAL LIFE
    • Game level design: can a player with given max jump widths
      cross a platform set?
    • Network reliability: can packets hop router-to-router given
      per-router max TTL hops?
    • Supply-chain feasibility: max pallet transfers per depot.
=====================================================================
"""

from __future__ import annotations
import time
import tracemalloc
from typing import List

# =====================================================================
# APPROACH 1 : Greedy furthest-reach (best)
# =====================================================================
class SolutionOne:
    """
    Purpose : Decide whether the last index is reachable from index 0.
    Inputs  : nums — list of ints, nums[i] = max forward jump length.
    Output  : bool — True if index len(nums)-1 is reachable.
    """

    def canJump(self, nums: List[int]) -> bool:
        reach = 0  # furthest index reachable so far (frontier)

        # Sweep; the pointer i must never overtake the frontier.
        for i in range(len(nums)):
            if i > reach:
                return False                 # index i is unreachable → stuck
            reach = max(reach, i + nums[i])  # grow frontier from here
            if reach >= len(nums) - 1:
                return True                  # last index covered
        return True

# =====================================================================
# APPROACH 2 : DP reachable[] frontier (O(n²), for contrast)
# =====================================================================
class SolutionTwo:
    """
    Purpose : Decide whether the last index is reachable from index 0.
    Inputs  : nums — list of ints, nums[i] = max forward jump length.
    Output  : bool — True if index len(nums)-1 is reachable.
    """

    def canJump(self, nums: List[int]) -> bool:
        reachable = [False] * len(nums)
        reachable[0] = True  # start cell is trivially reachable

        for i in range(len(nums)):
            if not reachable[i]:
                continue                     # dead cell: nothing propagates
            # Stamp every landing cell this jump offers as reachable.
            for j in range(i + 1, min(i + nums[i], len(nums) - 1) + 1):
                reachable[j] = True
        return reachable[-1]


# =====================================================================
# BENCHMARK — both approaches, time + peak memory per approach
# =====================================================================
if __name__ == "__main__":
    # Official examples + a zero-gap trap + never-leaves-start case.
    tests: List[List[int]] = [
        [2, 3, 1, 1, 4],   # → True  (2→3→4)
        [3, 2, 1, 0, 4],   # → False (stuck at index 3)
        [2, 0, 0],         # → True  (lands exactly on the end)
        [0, 1],            # → False (never leaves index 0)
    ]
    expected = [True, False, True, False]

    def run_all(solve) -> tuple[bool, float, int]:
        """Time one solver over every test; return (allPass, µs, peakBytes)."""
        tracemalloc.start()
        t0 = time.perf_counter()
        ok = all(solve(t) == e for t, e in zip(tests, expected))
        t1 = time.perf_counter()
        _, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        return ok, (t1 - t0) * 1e6, peak

    ok1, us1, peak1 = run_all(lambda t: SolutionOne().canJump(t))
    ok2, us2, peak2 = run_all(lambda t: SolutionTwo().canJump(t))

    print(f"Approach 1 (greedy reach) : {'PASS' if ok1 else 'FAIL'}  time = {us1:.2f} µs  peak-mem = {peak1} bytes")
    print(f"Approach 2 (DP frontier)   : {'PASS' if ok2 else 'FAIL'}  time = {us2:.2f} µs  peak-mem = {peak2} bytes")
    print("PASS : both approaches agree on every test case."
          if ok1 and ok2
          else "FAIL : at least one approach is wrong.")

File: neetcode150/15_greedy/0055_jump_game.cpp
/*
 * =====================================================================
 * LeetCode 55 : Jump Game                                           (Medium)
 * https://leetcode.com/problems/jump-game/
 * Category   : Greedy
 * ---------------------------------------------------------------------
 * PROBLEM
 *   You start at index 0 of `nums`. From index i you can jump at most
 *   nums[i] steps FORWARD. Return true if you can reach the last
 *   index.
 * ---------------------------------------------------------------------
 * INTUITION
 *   We never need to know WHICH jumps to take — only how far we can
 *   get. Track a single scalar: the furthest reachable index. When
 *   the sweep pointer passes the reach, the array is not reachable.
 *   Greedy because a larger reach strictly dominates a smaller one:
 *   every target that a worse path reaches, our best path reaches too
 *   (exchange argument — swap in our reach for theirs at every step).
 * ---------------------------------------------------------------------
 * APPROACH 1 — Greedy furthest-reach (BEST, O(1) space)
 *     reach = max(reach, i + nums[i])
 *     if i > reach → index i is unreachable → return false.
 *   Time  Complexity : O(n)   Space Complexity : O(1)
 *   WHY it is correct: reachable[i] is decided by the max over all
 *   j < i of (j + nums[j] >= i). The single number `reach` encodes
 *   exactly that max — it is a sufficient statistic. No DP table
 *   needed because "reachable" is monotone: positions ≤ reach are
 *   ALL reachable, so boundary checking is enough.
 * ---------------------------------------------------------------------
 * APPROACH 2 — Dynamic programming O(n²) (shows why greedy wins)
 *   reachable[0] = true; for every reachable i, mark all of
 *   (i, i+nums[i]] as reachable.
 *   Time  Complexity : O(n²)   Space Complexity : O(n)
 *   Correct but wasteful: it re-derives the same frontier again and
 *   again. Greedy compresses the whole reachable frontier into ONE
 *   integer — the DP table's information is redundant because the
 *   frontier is a single interval [0, reach].
 * ---------------------------------------------------------------------
 * DEEP DIVE — why "furthest reach" greedy cannot fail
 *   Exchange argument: let O be an optimal (winning) path reaching
 *   the end, and let G be our greedy path. Claim: after k moves, the
 *   greedy position is ≥ the optimal position for EVERY k. Induction:
 *   greedy position g_k ≥ o_k, and our jump can reach g_k + nums[g_k]
 *   ≥ o_k + nums[o_k] ≥ o_{k+1}. So greedily we are always at least as
 *   far as any rival — hence we can always imitate and beat O. This
 *   monotone-domination structure is the signature of a safe greedy.
 * =====================================================================
 */

#include <iostream>       // std::cout — benchmark output
#include <vector>         // std::vector — input array
#include <string>         // std::string — PASS/FAIL messages
#include <chrono>         // std::chrono — high-resolution timing
#include <sys/resource.h> // getrusage — peak RSS measurement

/* =====================================================================
 * APPROACH 1 : Greedy furthest-reach
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Decide whether the last index is reachable.
     * Inputs  : nums — vector of integers, nums[i] = max forward jump.
     * Output  : bool — true if index n-1 is reachable from index 0.
     */
    bool canJump(const std::vector<int>& nums) {
        // Invariant: `reach` = furthest index reachable so far.
        int reach = 0;

        // Sweep left to right; each index either extends the frontier
        // or is discovered unreachable (sweep pointer overtakes it).
        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            if (i > reach) return false;       // cannot even land on i
            if (i + nums[i] > reach) {
                reach = i + nums[i];           // frontier grows rightwards
            }
            if (reach >= static_cast<int>(nums.size()) - 1) {
                return true;                   // last index covered — done
            }
        }
        return true;
    }
};

/* =====================================================================
 * APPROACH 2 : DP with a reachable[] frontier (O(n²), for contrast)
 * ===================================================================== */
class Solution_2 {
public:
    /*
     * Purpose : Decide whether the last index is reachable.
     * Inputs  : nums — vector of integers, nums[i] = max forward jump.
     * Output  : bool — true if index n-1 is reachable from index 0.
     */
    bool canJump(const std::vector<int>& nums) {
        // reachable[i] = "index i can be reached". Re-derives the same
        // interval frontier as the greedy, but stores n booleans.
        std::vector<bool> reachable(nums.size(), false);
        reachable[0] = true; // start cell is trivially reachable

        for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
            if (!reachable[i]) continue;       // dead cell — no propagation

            // Stamp every landing cell this jump offers as reachable.
            int limit = i + nums[i];
            if (limit >= static_cast<int>(nums.size())) limit = nums.size() - 1;
            for (int j = i + 1; j <= limit; ++j) {
                reachable[j] = true;
            }
        }
        return reachable.back();
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches on 3 tests, time + peak RAM
 * ===================================================================== */
int main() {
    // Official examples + a tricky zero-gap case.
    const std::vector<std::vector<int> > tests = {
        { 2, 3, 1, 1, 4 },        // → true  (2→3→4)
        { 3, 2, 1, 0, 4 },        // → false (stuck at index 3)
        { 2, 0, 0 },              // → true  (2 lands exactly on the end)
        { 0, 1 }                  // → false (never leave index 0)
    };
    const std::vector<bool> expected = { true, false, true, false };

    auto run_all = [&](int which, bool (*solve)(const std::vector<int>&)) {
        struct rusage before, after;
        getrusage(RUSAGE_SELF, &before);
        auto t0 = std::chrono::high_resolution_clock::now();

        bool ok = true;
        for (int t = 0; t < static_cast<int>(tests.size()); ++t) {
            if (solve(tests[t]) != expected[t]) ok = false;
        }

        auto t1 = std::chrono::high_resolution_clock::now();
        getrusage(RUSAGE_SELF, &after);
        double us = std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count() / 1000.0;
        long mem_kb = after.ru_maxrss - before.ru_maxrss; // delta, KB

        std::cout << "Approach " << which
                  << " : " << (ok ? "PASS" : "FAIL")
                  << "  time = " << us << " µs  peak-ram-delta = " << mem_kb << " KB\n";
        return ok;
    };

    bool a1 = run_all(1, [](const std::vector<int>& v) { return Solution_1().canJump(v); });
    bool a2 = run_all(2, [](const std::vector<int>& v) { return Solution_2().canJump(v); });
    std::cout << (a1 && a2 ? "PASS : both approaches agree on every test case."
                           : "FAIL : at least one approach is wrong.") << "\n";
    return 0;
}
